contracts/evm/contracts/xcall/CallService.sol

```diff
@@ -116,122 +111,221 @@ contract CallService is IBSH, ICallService, IFeeManage, Initializable {
-        string[] memory sources,
-        string[] memory destinations
-    ) internal returns (
-        uint256
-    ) {
-        // check if caller is a contract or rollback data is null in case of EOA
-        require(msg.sender.code.length > 0 || _rollback.length == 0, "RollbackNotPossible");
-
-        // check size of payloads to avoid abusing
-        require(_rollback.length <= MAX_ROLLBACK_SIZE, "MaxRollbackSizeExceeded");
-
-        bool needResponse = _rollback.length > 0;
-        (string memory netTo, string memory dstAccount) = _to.parseNetworkAddress();
-        string memory from = nid.networkAddress(msg.sender.toString());
+        bytes memory _data
+    ) public payable returns (uint256) {
+        address caller = msg.sender;
+        Types.XCallEnvelope memory envelope = _data.decodeXCallEnvelope();
         uint256 sn = getNextSn();
-        int256 msgSn = 0;
-        if (needResponse) {
-            requests[sn] = Types.CallRequest(msg.sender, netTo, sources, _rollback, false);
-            msgSn = int256(sn);
-        }
-        Types.CSMessageRequest memory reqMsg = Types.CSMessageRequest(
-            from, dstAccount, sn, needResponse, _data, destinations);
-        bytes memory _msg = reqMsg.encodeCSMessageRequest();
+        Types.ProcessResult memory result = preProcessMessage(
+            sn,
+            _to,
+            envelope
+        );
+        string memory from = nid.networkAddress(caller.toString());
+
+        (string memory netTo, string memory dstAccount) = _to
+            .parseNetworkAddress();
+
+        emit CallMessageReq(from, _to, sn, envelope.messageType, result.data);
+        Types.CSMessageRequest memory req = Types.CSMessageRequest(
+            from,
+            dstAccount,
+            sn,
+            envelope.messageType,
+            result.data,
+            envelope.destinations
+        );
+
+        bytes memory _msg = req.encodeCSMessageRequest();
         require(_msg.length <= MAX_DATA_SIZE, "MaxDataSizeExceeded");
 
+        uint256 sendSn = result.needResponse ? sn : 0;
+
+        sendMessage(
+            envelope.sources,
+            netTo,
+            Types.CS_REQUEST,
+            int(sendSn),
+            _msg
+        );
+        claimProtocolFee();
+        emit CallMessageSent(caller, _to, sn);
+        return sn;
+    }
+
+    function sendMessage(
+        string[] memory sources,
+        string memory netTo,
+        int msgType,
+        int256 sn,
+        bytes memory data
+    ) private {
         if (sources.length == 0) {
             address conn = defaultConnections[netTo];
             require(conn != address(0), "NoDefaultConnection");
-            uint256 requiredFee = IConnection(conn).getFee(netTo, needResponse);
-            sendBTPMessage(conn, requiredFee, netTo, Types.CS_REQUEST, msgSn, _msg);
+            uint256 requiredFee = _getFee(conn, netTo, sn);
+            sendBTPMessage(conn, requiredFee, netTo, msgType, sn, data);
```

Comment:
Can we rename sendBTPMessage to _sendMessage or sendToConnection or  similar in this PR?

Reply:
renaming this would be okay I guess, will rename it sendToConnection!